draft-ietf-jsonpath-base.md

```diff
@@ -1176,6 +1181,72 @@ The semantics of regular expressions are as defined in {{-iregexp}}.
 The logical AND, OR, and NOT operators have the normal semantics of Boolean algebra and
 obey its laws (see, for example, {{BOOLEAN-LAWS}}).
 
+##### Function Extensions {#fnex}
+{: unnumbered}
+
+[^unnumbered-bad]
+
+[^unnumbered-bad]: This should not be an unnumbered section, which it is
+    forced to be by the parent section being unnumbered too.
+    The overall structure where the parts of this subsection go to
+    needs to be decided; one part already is in the IANA
+    Considerations in {{iana-fnex}}
+{:source=" -- cabo"}
+
+Beyond the filter expression functionality defined in the preceding
+subsections, JSONPath defines an extension point that can be used to
+add filter expression functionality: "Function Extensions".
+
+A function extension defines a registered name (see {{iana-fnex}}) that
+can be applied to a sequence of zero or more arguments, resulting in a result.
+A function-expression stands for ("returns") a nodelist or a value.
+The arguments are filter expressions that are to be interpreted as
+nodelists or values.
+For each argument and for the result, the function extension defines
+the "kind" of the item, i.e., whether the item is a value or a
+nodelist ("nodes").
+
+~~~ abnf
+function-name           = function-name-first *function-name-char
+function-name-first     = LCALPHA
+function-name-char      = DIGIT / name-first / "_"
+LCALPHA                 = %x61-7A  ; "a".."z"
+
+
+function-expression     = function-name "(" [*(function-argument "," S) function-argument] ")"
+singular-path           =/ function-expression
+function-argument       = comparable
+~~~
+
+Syntactically, a function-expression can occur anywhere where a
+singular-path can occur (exist-expr, comparable, regex-expr).
+A function-expression that employs a function extension that returns a
+nodelist MUST return a singular node to be used in a comparable or a
+regex-expr.
+
+###### Function Extensions: length {#length}
+{: unnumbered}
+
+When defining an extension point in a protocol, it is generally a good
+idea to provide at least one extension that makes use of the extension
+point, to serve as an example and to provide motivation to implement
+the extension point.
+
+The "length" function extension provides a way to compute the length
+of an array and make that available for further processing in the
+filter expression:
+
+~~~ JSONPath
+$[?length(@.authors) >= 5]
+~~~
+
+Its only argument is a value (possibly taken from a singular path as
+in the example above).  The result also is a value; if the argument
+value is an array, the result is the number of elements in the array.
+[length-mismatch]
+
+[length-mismatch]: Define the case where the argument is not an array as well.
```

Comment:
It seems we have limited options if we are going to be consistent with the rest of the spec. We could either return a fixed value (e.g. `-1`) or an empty nodelist. The latter seems preferable as it's consistent with a singular path possibly being empty.

This makes me wonder whether we should make a function always return a nodelist and have the function spec say whether or not the returned nodelist is "singular" (meaning it has at most one member) instead of allowing nodelists of values to be returned.

Reply:
The main problem is that the node that has the length isn't in the input JSON value

Reply:
Of course. I was thinking "list of nodes" rather than _nodelist_.

Reply:
If we can avoid the function having to create nodes, so much the better. So I guess that limits return values to nodelists or values after all.

Reply:
I like the well known return value `-1` for the case of an invalid argument.

Reply:
I made a different proposal specifically for length(), but we do need to come up with a more general solution.